op-node/rollup/derive/channel_test.go

```diff
@@ -0,0 +1,85 @@
+package derive
+
+import (
+	"testing"
+
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/stretchr/testify/require"
+)
+
+type frameValidityTC struct {
+	name      string
+	frames    []Frame
+	shouldErr []bool
+	sizes     []uint64
+}
+
+func (tc *frameValidityTC) Run(t *testing.T) {
+	id := [16]byte{0xff}
+	block := eth.L1BlockRef{}
+	ch := NewChannel(id, block)
+
+	if len(tc.frames) != len(tc.shouldErr) || len(tc.frames) != len(tc.sizes) {
+		t.Errorf("lengths should be the same. frames: %d, shouldErr: %d, sizes: %d", len(tc.frames), len(tc.shouldErr), len(tc.sizes))
+	}
+
+	for i, frame := range tc.frames {
+		err := ch.AddFrame(frame, block)
+		if tc.shouldErr[i] {
+			require.NotNil(t, err)
+		} else {
+			require.Nil(t, err)
+		}
+		require.Equal(t, tc.sizes[i], ch.Size())
+	}
+}
+
+// TestFrameValidity inserts a list of frames into the channel. It checks if an error
+// should be returned by `AddFrame` as well as checking the size of the channel.
+func TestFrameValidity(t *testing.T) {
+	id := [16]byte{0xff}
+	testCases := []frameValidityTC{
+		{
+			name:      "wrong channel",
+			frames:    []Frame{{ID: [16]byte{0xee}}},
+			shouldErr: []bool{true},
+			sizes:     []uint64{0},
+		},
+		{
+			name: "double close",
+			frames: []Frame{
+				{ID: id, FrameNumber: 2, IsLast: true, Data: []byte("four")},
+				{ID: id, FrameNumber: 1, IsLast: true}},
+			shouldErr: []bool{false, true},
+			sizes:     []uint64{204, 204},
+		},
+		{
+			name: "duplicate frame",
+			frames: []Frame{
+				{ID: id, FrameNumber: 2, Data: []byte("four")},
+				{ID: id, FrameNumber: 2, Data: []byte("seven__")}},
+			shouldErr: []bool{false, true},
```

Comment:
Is a duplicate frame an error or is the frame supposed to just be dropped? Will this prevent further parsing of the batch data?

Reply:
Frame is supposed to be dropped

Reply:
`IngestFrame` (which uses `AddFrame` is an infallible function. It simply logs the error (+ also skips the CB prune, but this is fine b/c the size would not have grow if err != nil & the CB would have been in compliance with it's size limit prior to entering IngestFrame)

https://github.com/ethereum-optimism/optimism/blob/e081a971958157191ea9b5ad8264b8c12d58cd7c/op-node/rollup/derive/channel_bank.go#L96-L100